### DISCUSSION TITLE
Close App vs Close Window

### DISCUSSION THREAD

**Discussion by jrock2004** [hyprwm/Hyprland#2036] [Q&A]
So I binded  to run the command `killactive` but that seems to just close the window and not actually the app. If I do a ps aux I still see the app active. How can I actually close the app completely?

```
...
bind = SUPER, Q, killactive,
...
```

Thanks

**Comment by vaxerski** [comment]
ask the app to not minimize to tray on close?

or just `killall -9 <appname>` I guess

**Reply by jrock2004** [reply]
So I guess for now the best way to use my mouse to exit the app instead of minimize to tray or close window

**Reply by vaxerski** [reply]
killactive is poorly named. It sends a "please close" signal to the app. If it handles that by minimizing to tray, it's the problem of the app, not hyprland

**Reply by jrock2004** [reply]
Yeah, I assumed it actually killed the app. So what you are thinking is that since no other windows is open its up to the app to kill itself?

For example Firefox will does not close when I kill it, so we need to hope that some day Firefox will fix this?

**Reply by vaxerski** [reply]
well the app should just "exit" gracefuly

firefox for me quits when I close it with a keybind.

If you really need to kill it, you can either use `hyprctl kill` (works like xkill) or make a script that checks `hyprctl activewindow` and runs kill on its PID

**Reply by jrock2004** [reply]
So Firefox closes for sure, but if I restart my computer and open Firefox it tells me it crashed and if I want to re-open tabs. Maybe something on my machine

**Reply by n3rdw1z4rd** [reply]
I am experiencing the same behavior with Chromium, but not every time. I have figured out that when I am shutting down my system, I use the keybind that equates to `killactive` until all windows are gone, then I issue the keybind that equates to `exit` Hyprland. I found out that I need to wait a minute before shutting down Hyprland and it works fine. This may be because Chromium is saving state or something and I'm killing Hyprland before it finishes...

This is just a speculation, not based on any evidence. But I am curious to know if my speculation is correct...

Could try to add a short timer to the `exit` keybind to quit Hyprland to allow for apps to save state...

**Reply by a-amr** [reply]
there are many types of kiill signals you can chose i belive you need to send a soft signal to close the app not hard close of force to close.
 you can use kill -l to see the list 

hyprctl activewindow |grep pid |tr -d 'pid:'|xargs kill
i use this in my waybar to kill the curent app on middle click just to be able to close any app if i only use mouse

**Reply by n3rdw1z4rd** [reply]
Thanks @a-amr, I've switched away from Hyprland for now, but I still do dabble, and will try your suggestion - thanks again :)

**Comment by notalwaysbum** [comment]
Don't know if back when question was asked that was the case, but it appears there's already `forcekillactive` dispatcher which handles complete shutdown of apps. More on [wiki](https://wiki.hypr.land/Configuring/Dispatchers/#list-of-dispatchers).

I binded it like:
bind = $mainMod, C, killactive
bind = $mainMod SHIFT, C, forcekillactive

Works like a charm.
